src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceCodeRefactoringProvider.MoveFileCodeAction.cs

```diff
@@ -0,0 +1,162 @@
+            {
+                if (parts.Length == 0)
+                {
+                    return ImmutableArray.Create(currentFolder);
+                }
+
+                var partsArray = parts.ToArray();
+                var candidates = Enumerable.Range(1, parts.Length)
+                    .Select(i => (foldername: string.Join(".", partsArray, 0, i), index: i))     // index is the number of items in parts used to construct key
+                    .ToImmutableDictionary(t => t.foldername, t => t.index, PathUtilities.Comparer);
+
+                var subFolders = currentFolderInfo.Folders;
+
+                var builder = ArrayBuilder<ImmutableArray<string>>.GetInstance();
+                foreach ((var folderName, var index) in candidates)
+                {
+                    if (subFolders.TryGetValue(folderName, out var matchingFolderInfo))
+                    {
+                        var newParts = index >= parts.Length
+                            ? ImmutableArray<string>.Empty
+                            : ImmutableArray.Create(parts, index, parts.Length - index);
+                        var newCurrentFolder = currentFolder.Add(matchingFolderInfo.Name);
+                        builder.AddRange(FindCandidateFolders(matchingFolderInfo, newParts, newCurrentFolder));
+                    }
+                }
+
+                var defaultPathBasedOnCurrentFolder = currentFolder.AddRange(parts);
+                if (builder.All(folders => !folders.SequenceEqual(defaultPathBasedOnCurrentFolder, PathUtilities.Comparer)))
+                {
+                    builder.Add(defaultPathBasedOnCurrentFolder);
+                }
+
+                return builder.ToImmutableAndFree();
+            }
+
+            private class FolderInfo
+            {
+                Dictionary<string, FolderInfo> _folders;
+
+                public string Name { get; }
+
+                public FolderInfo Parent { get; }
+
+                public IReadOnlyDictionary<string, FolderInfo> Folders => _folders;
+
+                private FolderInfo(string name, FolderInfo parent)
+                {
+                    Name = name;
+                    Parent = parent;
+                    _folders = new Dictionary<string, FolderInfo>(StringComparer.Ordinal);
+                }
+
+                public void AddFolder(ImmutableArray<string> folder)
+                {
+                    if (folder.IsDefaultOrEmpty)
+                    {
+                        return;
+                    }
+
+                    var firstFolder = folder[0];
+                    if (!_folders.TryGetValue(firstFolder, out var firstFolderInfo))
+                    {
+                        firstFolderInfo = new FolderInfo(firstFolder, this);
+                        _folders[firstFolder] = firstFolderInfo;
+                    }
+
+                    firstFolderInfo.AddFolder(ImmutableArray.CreateRange(folder.Skip(1)));
```

Comment:
this is N^2 or N^3.  that said, it may not matter with the number of folders in a project.

Reply:
I don't expect a project to have a large number of folders. And in case there were many, I don't expect this search to go very far (i.e. no deeply matching folders). But I will try to make this more efficient.

Reply:
you may not have many folders, but you may have many documents.  And you do the work each time for each document.  You might want to cache the folders you've seen and then just quickly check that when you look at each new document.  That way most of them would be no-ops, and only teh first document in each new folder chain actually did something.